File: two_pass_assembler/two-pass-assembler.py
#!/usr/bin/env python3

import sys
import re

class TwoPassAssembler:
    opcodes = { "MVI_A" : "3E",
                "LDA" : "3A",
                "STA" : "32",
                "MOV_B_A" : "47",
                "MVI_A" : "3E",
                "ADD_B" : "80",
                "ADI" : "C6" ,
                "HLT" : "76"}
    def __init__(self, filename, start_addr):
        self.filename = filename
        self.start_addr = start_addr
        self.table = {}
        self.labels = {}
        self.end_addr = start_addr

    def symbol_parse(self):
        prog = open(self.filename, "r")
        lines = prog.readlines()
        prog.close()
        addr = self.start_addr
        prev = ""
        for line in lines:
            lab = re.findall(r"([0-9a-zA-Z_]+)\s*:", line)
            labi = 0
            words = re.findall(r"[0-9a-zA-Z_]+", line)
            for word in words:
                if word in self.labels:
                    word = "%.4X"%self.labels[word]
                if prev in {"STA", "LDA", "JMP", "CALL"} and re.match(r"[0-9a-fA-F]{4}", word):
                    byte2, byte1 = word[:2], word[2:]
                    self.table[addr] = byte1
                    addr = (addr + 1)&0xFFFF
                    self.table[addr] = byte2
                    addr = (addr + 1)&0xFFFF
                elif labi < len(lab) and word == lab[labi]:
                    self.labels[word] = addr
                    labi += 1
                else:
                    self.table[addr] = word
                    addr = (addr + 1)&0xFFFF
                prev = word
        self.end_addr = addr

    def to_machine_code(self):
        machine_code = {}
        addr = self.start_addr
        while addr != self.end_addr:
            tok = self.table[addr]
            if tok in __class__.opcodes:
                tok = __class__.opcodes[tok]
            machine_code[addr] = tok
            addr = (addr + 1)&0xFFFF
        return machine_code

start_addr = int(sys.argv[2], 16)
if start_addr < 0x0000 or start_addr > 0xFFFF:
    print("Error: invalid starting address for 8085")
    sys.exit(-1)
filename = sys.argv[1]

assembler = TwoPassAssembler(filename, start_addr)
assembler.symbol_parse()
machine_code = assembler.to_machine_code()

addr = assembler.start_addr
print("Address\t\tPnemonic\tMachine Code")
while addr != assembler.end_addr:
    print("0x%.4X\t\t%s\t\t%s"%(addr, assembler.table[addr], machine_code[addr]))
    addr = (addr + 1)&0xFFFF

File: two_pass_assembler/two-pass-assembler.cpp
#include<iostream>
#include <fstream>
#include<streambuf>
#include<iomanip>
#include<unordered_map>
#include<string>
#include<regex>

using namespace std;

string int_to_hex(int i){
    std::stringstream stream;
    stream << std::setfill('0') << std::setw(4) << std::hex << i;
    return stream.str();
}

class TwoPassAssembler{
public:
    unordered_map<string, string> opcodes;
    string filename;
    int start_addr, end_addr;
    unordered_map<int, string> table;
    unordered_map<string, int> labels;

    TwoPassAssembler(string fname, int saddr) : 
    filename(fname), start_addr(saddr), end_addr(-1) {
        opcodes["MVI_A"] = "3E";
        opcodes["LDA"] = "3A";
        opcodes["STA"] = "32";
        opcodes["MOV_B_A"] = "47";
        opcodes["MVI_A"] = "3E";
        opcodes["ADD_B"] = "80";
        opcodes["ADI"] = "C6";
    }
    void symbol_parse(void);
    unordered_map<int, string> to_machine_code(void);
};

void TwoPassAssembler::symbol_parse(void){
    std::ifstream prog(filename);
    std::string lines;
    prog.seekg(0, std::ios::end);
    lines.reserve(prog.tellg());
    prog.seekg(0, std::ios::beg);
    lines.assign((std::istreambuf_iterator<char>(prog)),
               std::istreambuf_iterator<char>());
    prog.close();

    regex wordreg("([0-9a-zA-Z_]+)");
    regex labelreg("([0-9a-zA-Z_]+)\\s*:");
    int addr = start_addr;
    string prev = "";
    auto word = sregex_iterator(lines.begin(), lines.end(), wordreg);
    auto label = sregex_iterator(lines.begin(), lines.end(), labelreg);
    auto word_end = sregex_iterator(), label_end = sregex_iterator();
    vector<string> Addr_tokens {"STA", "LDA", "JMP", "CALL"};
    smatch extract;
    string slabel = "";
    if (label != label_end){
        slabel = label->str();
        regex_search(slabel, extract, labelreg);
        slabel = extract[1];
    }
    for(sregex_iterator wd = word; wd != word_end; wd++){
        string Word = wd->str();
        if (labels.find(Word) != labels.end()){
            Word = int_to_hex(labels[Word]);
        }
        if (find(Addr_tokens.begin(), Addr_tokens.end(), prev) != Addr_tokens.end()
            && regex_match(Word, regex("([0-9a-fA-F]{4})"))){
            string byte1 = Word.substr(2, 2), byte2 = Word.substr(0, 2);
            table[addr++] = byte1;
            table[addr++] = byte2;
        }else if (label != label_end && slabel == Word){
            labels[Word] = addr;
            label++;
            if (label != label_end){
                slabel = label->str();
                regex_search(slabel, extract, labelreg);
                slabel = extract[1];
            }
        }else {
            table[addr++] = Word;
        }
        prev = Word;
    }
    end_addr = addr;
}

unordered_map<int, string> TwoPassAssembler::to_machine_code(){
    unordered_map<int, string> machine_code;
    for(int addr = start_addr; addr < end_addr; addr++){
        string tok = table[addr];
        if (opcodes.find(tok) != opcodes.end())
            tok = opcodes[tok];
        machine_code[addr] = tok;
    }
    return machine_code;
}

int main(int argc, char *argv[]){
    string filename(argv[1]);
    string saddr(argv[2]);
    saddr = "0x" + saddr;
    int start_addr = stoul(saddr, nullptr, 16);
    
    TwoPassAssembler assembler(filename, start_addr);
    assembler.symbol_parse();
    unordered_map<int, string> machine_code;
    machine_code = assembler.to_machine_code();

    cout << "Address\t\tPnemonic\tMachine Code" << endl;
    for (int addr=assembler.start_addr; addr<assembler.end_addr; addr++){
        cout << hex << "0x" << addr << "\t\t";
        cout << assembler.table[addr] << "\t\t" << machine_code[addr] << endl;
    }
    return 0;
}